File: RNDO/IntelEdison/GyroRNDOSender/src/RNDOTransmitClient.cpp
//----------------------------------------------------------------------------
//
//  $Workfile: RNDOTransmitClent.cpp$
//
//  $Revision: X$
//
//  Project:    Robot Network Data Objects
//
//                            Copyright (c) 2016
//                               James A Wright
//                            All Rights Reserved
//
//  Modification History:
//  $Log:
//  $
//
//----------------------------------------------------------------------------
#include <iostream>
#include<sys/socket.h>    //socket
#include<arpa/inet.h> //inet_addr
#include "RNDOTransmitClient.h"

using namespace std;

RNDOTransmitClient::RNDOTransmitClient(string name)
{
	mName = name;
//	mTCPIPPortID = ConfigFile.returnTCPIPPort(name, "Receive");
	mTCPIPPortID = 4543;
	mTransmitCount = 0;
}

//--------------------------------------------------------------------
// Purpose:
//     Start the threads
//
// Notes:
//     None.
//--------------------------------------------------------------------
void RNDOTransmitClient::Start(void)
{
	mRunThread = true;
	pthread_create(&mTCPIPThread, 0, &RNDOTransmitClient::callMemberFunction, this);
}

//--------------------------------------------------------------------
// Purpose:
//     Stop the threads
//
// Notes:
//     None.
//--------------------------------------------------------------------
void RNDOTransmitClient::Stop(void)
{
	mRunThread = false;
}

//--------------------------------------------------------------------
// Purpose:
//
//
// Notes:
//     None.
//--------------------------------------------------------------------
RNDOMessage* RNDOTransmitClient::GetRNDOMessage(UInt16 id, UInt16 robot)
{
	RNDOMessage* message = new RNDOMessage(id,robot);

	for (int index = 0; index < (int) mRNDOMessages.size(); index++)
	{
		if (mRNDOMessages[index]->GetHeader() == message->GetHeader())
		{
			return mRNDOMessages[index];
		}
	}

	mRNDOMessages.push_back(message);
	return message;
}

//--------------------------------------------------------------------
// Purpose:
//
//
// Notes:
//     None.
//--------------------------------------------------------------------
RNDOMessage* RNDOTransmitClient::GetRNDOMessage(UInt16 id)
{
	return GetRNDOMessage(id,0);
}

//----------------------------------------------------------------------------
//  Purpose:
//      the thread that does the work
//
//  Notes:
//      None
//
//----------------------------------------------------------------------------
void* RNDOTransmitClient::RunTCPIPThread(void)
{
    int sock;
    struct sockaddr_in server;

    server.sin_addr.s_addr = inet_addr("192.168.143.193");
    server.sin_family = AF_INET;
    server.sin_port = htons( 4543 );

	while(true == mRunThread)
	{
		try
		{
		    sock = socket(AF_INET , SOCK_STREAM , 0);
		    if (sock == -1)
		    {
		        cout<<"Could not create socket"<<endl;
		    }
		    cout<<"Socket created"<<endl;

		    //Connect to remote server
		    int connected = connect(sock , (struct sockaddr *)&server , sizeof(server));

		    cout<<"Connected"<<connected<<endl;

			while ((true == mRunThread)&&(connected>=0))
			{
				try
				{
					UInt32 dataSize = (UInt32)((RNDOMessage::MAX_FULL_MESSAGE_SIZE * mRNDOMessages.size()));
					byte data[dataSize];

					for(int index=0;index<(int)mRNDOMessages.size();index++)
					{
						mRNDOMessages[index]->GetByteArray(data, (index * RNDOMessage::MAX_FULL_MESSAGE_SIZE));
					}

					send(sock , data , dataSize , 0);
					/*
					cout<<"send:"<<dataSize<<":";

					for(int index=0;index<dataSize;index++)
					{
						cout<<(int)data[index]<<" ";
					}
				    cout<<endl;
					*/
					//mTransmitTimeSpan = DateTime.Now.Subtract(transmitTime);
					//transmitTime = DateTime.Now;
					mTransmitCount++;
				}
				catch(...)
				{
				}
				usleep(8000);
			}
		}
		catch(...)
		{
		}
	}
	return NULL;
}

File: RNDO/Windows/RNDO/RNDOClient/RNDOTransmitClient.cs
//----------------------------------------------------------------------------
//
//  $Workfile: RNDOTransmitClient.cs$
//
//  $Revision: X$
//
//  Project:    Robot Network Data Objects
//
//                            Copyright (c) 2016
//                               James A Wright
//                            All Rights Reserved
//
//  Modification History:
//  $Log:
//  $
//
//----------------------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net.Sockets;
using System.Threading;
using System.Net;
using RNDO;

namespace RNDOClientInterface
{
    //----------------------------------------------------------------------------
    //  Class Declarations
    //----------------------------------------------------------------------------
    //
    // Class Name: RNDOTransmitClient
    // 
    // Purpose:
    //      This class handles sending messages to the server
    //
    //----------------------------------------------------------------------------
    public class RNDOTransmitClient
    {
        public string mName;
        bool mRunThread = true;
        Thread mTCPIPThread;
        UInt32 mTCPIPPortID;
        List<RNDOMessage> mRNDOMessages = new List<RNDOMessage>();
        public UInt32 mTransmitCount = 0;
        public TimeSpan mTransmitTimeSpan = new TimeSpan();

        //----------------------------------------------------------------------------
        //  Purpose:
        //      Constructor
        //
        //  Notes:
        //      None
        //
        //----------------------------------------------------------------------------
        public RNDOTransmitClient(string name)
        {
            mName = name;
            mTCPIPPortID = ConfigFile.returnTCPIPPort(name, "Receive");
        }

        //--------------------------------------------------------------------
        // Purpose:
        //     Start the threads
        //
        // Notes:
        //     None.
        //--------------------------------------------------------------------
        public void Start()
        {
            mRunThread = true;
            mTCPIPThread = new Thread(new ThreadStart(RunTCPIPThread));
            mTCPIPThread.Name = string.Format("Handle TCP/IP " + mName);
            mTCPIPThread.Start();
        }

        //--------------------------------------------------------------------
        // Purpose:
        //     Stop the threads
        //
        // Notes:
        //     None.
        //--------------------------------------------------------------------
        public void Stop()
        {
            mRunThread = false;
        }

        //--------------------------------------------------------------------
        // Purpose:
        //     
        //
        // Notes:
        //     None.
        //--------------------------------------------------------------------
        public RNDOMessage GetRNDOMessage(UInt16 id)
        {
            RNDOMessage message = new RNDOMessage(id);

            for (int index = 0; index < mRNDOMessages.Count; index++)
            {
                if (mRNDOMessages[index].GetHeader() == message.GetHeader())
                {
                    return mRNDOMessages[index];
                }
            }

            mRNDOMessages.Add(message);

            return message;
        }

        //--------------------------------------------------------------------
        // Purpose:
        //     
        //
        // Notes:
        //     None.
        //--------------------------------------------------------------------
        public RNDOMessage GetRNDOMessage(UInt16 id, UInt16 robot)
        {
            RNDOMessage message = new RNDOMessage(id, robot);

            for (int index = 0; index < mRNDOMessages.Count; index++)
            {
                if (mRNDOMessages[index].GetHeader() == message.GetHeader())
                {
                    return mRNDOMessages[index];
                }
            }

            mRNDOMessages.Add(message);

            return message;
        }

        //----------------------------------------------------------------------------
        //  Purpose:
        //      the thread that does the work
        //
        //  Notes:
        //      None
        //
        //----------------------------------------------------------------------------
        private void RunTCPIPThread()
        {
            DateTime transmitTime = DateTime.Now;
            IPEndPoint localEndPoint = new IPEndPoint(IPAddress.Parse(ConfigFile.mRNDOIP), (int)mTCPIPPortID);

            while (true == mRunThread)
            {
                try
                {
                    Socket sender = new Socket(AddressFamily.InterNetwork,
                                                    SocketType.Stream,
                                                    ProtocolType.Tcp);

                    sender.Connect(localEndPoint);
                    sender.ReceiveTimeout = 6;
                    sender.SendTimeout = 6;

                    while ((true == mRunThread)&&(true == sender.Connected))
                    {
                        try
                        {
                            UInt32 dataSize = (UInt32)((RNDOMessage.MAX_FULL_MESSAGE_SIZE * mRNDOMessages.Count));
                            byte[] data = new byte[dataSize];

                            for(int index=0;index<mRNDOMessages.Count;index++)
                            {
                                mRNDOMessages[index].GetByteArray(data, (index * RNDOMessage.MAX_FULL_MESSAGE_SIZE));
                            }

                            sender.Send(data);
                            mTransmitTimeSpan = DateTime.Now.Subtract(transmitTime);
                            transmitTime = DateTime.Now;
                            mTransmitCount++;
                        }
                        catch(Exception)
                        {
                        }
                        Thread.Sleep(9);
                    }
                }
                catch(Exception)
                {
                }
                Thread.Sleep(9);
            }
        }
    }
}
